Migrate from Hilt (Dagger2) to Koin for Dependency Injection

Due to issues #4 originating from Hilt, the project is now being migrated to Koin. 
Koin is a light-weight DI framework written in Kotlin, and has been adopted by the developer community widely. 

### References

* https://blog.kotzilla.io/migrate-from-hilt-to-koin
* https://developer.android.com/build/migrate-to-ksp
* https://stackoverflow.com/a/77720024/13546426
* https://stackoverflow.com/a/71665221/13546426
* https://stackoverflow.com/a/77515662/13546426

This code compiles however once run, the following error happens where the generated files are not in the expected directory structure?

Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_Unwind_Resume" referenced by "/data/app/com.ml.shubham0204.docqa-DqcFSrNN4B3M38jFa8DzAA==/lib/x86/libhftokenizer.so"...
                                                                                                    	at java.lang.Runtime.loadLibrary0(Runtime.java:1016)
                                                                                                    	at java.lang.System.loadLibrary(System.java:1669)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.HFTokenizer.<clinit>(HFTokenizer.kt:51)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invokeSuspend(SentenceEmbedding.kt:24) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invoke(Unknown Source:8) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invoke(Unknown Source:4) 
                                                                                                    	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:78) 
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:167) 
                                                                                                    	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding.init(SentenceEmbedding.kt:23) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding.init$default(SentenceEmbedding.kt:18) 
                                                                                                    	at com.ml.shubham0204.docqa.domain.embeddings.SentenceEmbeddingProvider$1.invokeSuspend(SentenceEmbeddingProvider.kt:18) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 


Longer error output
-----------------------------
024-10-18 13:37:11.270 16523-16591 ubham0204.docq          com.ml.shubham0204.docqa             I    at void kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run() (CoroutineScheduler.kt:684)
2024-10-18 13:37:11.270 16523-16591 ubham0204.docq          com.ml.shubham0204.docqa             I  
2024-10-18 13:37:11.271 16523-16523 AndroidRuntime          com.ml.shubham0204.docqa             D  Shutting down VM
2024-10-18 13:37:11.271 16523-16523 AndroidRuntime          com.ml.shubham0204.docqa             E  FATAL EXCEPTION: main
                                                                                                    Process: com.ml.shubham0204.docqa, PID: 16523
                                                                                                    java.lang.NoClassDefFoundError: com.ml.shubham0204.sentence_embeddings.HFTokenizer
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invokeSuspend(SentenceEmbedding.kt:24)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invoke(Unknown Source:8)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invoke(Unknown Source:4)
                                                                                                    	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:78)
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:167)
                                                                                                    	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding.init(SentenceEmbedding.kt:23)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding.init$default(SentenceEmbedding.kt:18)
                                                                                                    	at com.ml.shubham0204.docqa.domain.embeddings.SentenceEmbeddingProvider$1.invokeSuspend(SentenceEmbeddingProvider.kt:18)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_Unwind_Resume" referenced by "/data/app/com.ml.shubham0204.docqa-DqcFSrNN4B3M38jFa8DzAA==/lib/x86/libhftokenizer.so"...
                                                                                                    	at java.lang.Runtime.loadLibrary0(Runtime.java:1016)
                                                                                                    	at java.lang.System.loadLibrary(System.java:1669)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.HFTokenizer.<clinit>(HFTokenizer.kt:51)
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invokeSuspend(SentenceEmbedding.kt:24) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invoke(Unknown Source:8) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding$init$2.invoke(Unknown Source:4) 
                                                                                                    	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:78) 
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:167) 
                                                                                                    	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding.init(SentenceEmbedding.kt:23) 
                                                                                                    	at com.ml.shubham0204.sentence_embeddings.SentenceEmbedding.init$default(SentenceEmbedding.kt:18) 
                                                                                                    	at com.ml.shubham0204.docqa.domain.embeddings.SentenceEmbeddingProvider$1.invokeSuspend(SentenceEmbeddingProvider.kt:18) 
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
2024-10-18 13:37:11.279 16523-16523 Process                 com.ml.shubham0204.docqa             I  Sending signal. PID: 16523 SIG: 9
2024-10-18 13:37:11.328  1996-2087  InputDispatcher         system_server                        E  channel 'b33a08e com.ml.shubham0204.docqa/com.ml.shubham0204.docqa.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!

--- a/app/build.gradle.kts
+++ b/app/build.gradle.kts
@@ -4,8 +4,7 @@ plugins {
     alias(libs.plugins.android.application)
     alias(libs.plugins.jetbrains.kotlin.android)
     alias(libs.plugins.compose.compiler)
-    kotlin("kapt")
-    id("com.google.dagger.hilt.android")
+    id("com.google.devtools.ksp")
 }
 
 val geminiKey: String = gradleLocalProperties(rootDir, providers).getProperty("geminiKey")
@@ -58,6 +57,17 @@ android {
             excludes += "META-INF/DEPENDENCIES"
         }
     }
+    applicationVariants.configureEach {
+        kotlin.sourceSets {
+            getByName(name) {
+                kotlin.srcDir("build/generated/ksp/$name/kotlin")
+            }
+        }
+    }
+}
+
+ksp {
+    arg("KOIN_CONFIG_CHECK","true")
 }
 
 dependencies {
@@ -94,9 +104,11 @@ dependencies {
     // https://github.com/jeziellago/compose-markdown
     implementation("com.github.jeziellago:compose-markdown:0.5.0")
 
-    implementation(libs.hilt.android)
-    kapt(libs.hilt.android.compiler)
-    implementation(libs.androidx.hilt.navigation.compose)
+    // Koin dependency injection
+    implementation(libs.koin.android)
+    implementation(libs.koin.annotations)
+    implementation(libs.koin.androidx.compose)
+    ksp(libs.koin.ksp.compiler)
 
     testImplementation(libs.junit)
     androidTestImplementation(libs.androidx.junit)

--- a/app/src/main/java/com/ml/shubham0204/docqa/DocQAApplication.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/DocQAApplication.kt
@@ -2,13 +2,19 @@ package com.ml.shubham0204.docqa
 
 import android.app.Application
 import com.ml.shubham0204.docqa.data.ObjectBoxStore
-import dagger.hilt.android.HiltAndroidApp
+import com.ml.shubham0204.docqa.di.AppModule
+import org.koin.android.ext.koin.androidContext
+import org.koin.core.context.startKoin
+import org.koin.ksp.generated.module
 
-@HiltAndroidApp
 class DocQAApplication : Application() {
 
     override fun onCreate() {
         super.onCreate()
+        startKoin {
+            androidContext(this@DocQAApplication)
+            modules(AppModule().module)
+        }
         ObjectBoxStore.init(this)
     }
 }

--- a/app/src/main/java/com/ml/shubham0204/docqa/MainActivity.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/MainActivity.kt
@@ -11,9 +11,7 @@ import androidx.navigation.compose.composable
 import androidx.navigation.compose.rememberNavController
 import com.ml.shubham0204.docqa.ui.screens.ChatScreen
 import com.ml.shubham0204.docqa.ui.screens.DocsScreen
-import dagger.hilt.android.AndroidEntryPoint
 
-@AndroidEntryPoint
 class MainActivity : ComponentActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {

--- a/app/src/main/java/com/ml/shubham0204/docqa/data/ChunksDB.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/data/ChunksDB.kt
@@ -1,5 +1,8 @@
 package com.ml.shubham0204.docqa.data
 
+import org.koin.core.annotation.Single
+
+@Single
 class ChunksDB {
 
     private val chunksBox = ObjectBoxStore.store.boxFor(Chunk::class.java)

--- a/app/src/main/java/com/ml/shubham0204/docqa/data/DocumentsDB.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/data/DocumentsDB.kt
@@ -5,7 +5,9 @@ import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.flowOn
+import org.koin.core.annotation.Single
 
+@Single
 class DocumentsDB {
 
     private val docsBox = ObjectBoxStore.store.boxFor(Document::class.java)

--- a/app/src/main/java/com/ml/shubham0204/docqa/di/AppModule.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/di/AppModule.kt
@@ -1,46 +1,7 @@
 package com.ml.shubham0204.docqa.di
 
-import android.app.Application
-import com.ml.shubham0204.docqa.data.ChunksDB
-import com.ml.shubham0204.docqa.data.DocumentsDB
-import com.ml.shubham0204.docqa.domain.embeddings.SentenceEmbeddingProvider
-import com.ml.shubham0204.docqa.domain.llm.GeminiRemoteAPI
-import dagger.Module
-import dagger.Provides
-import dagger.hilt.InstallIn
-import dagger.hilt.components.SingletonComponent
-import javax.inject.Singleton
+import org.koin.core.annotation.ComponentScan
+import org.koin.core.annotation.Module
 
-// AppModule provides dependencies that are to be injected by Hilt
-@Module
-@InstallIn(SingletonComponent::class)
-object AppModule {
-
-    // SingletonComponent ensures that instances survive
-    // across the application's lifespan
-    // @Singleton creates a single instance in the app's lifespan
-
-    @Provides
-    @Singleton
-    fun provideDocumentsDB(): DocumentsDB {
-        return DocumentsDB()
-    }
-
-    @Provides
-    @Singleton
-    fun provideChunksDB(): ChunksDB {
-        return ChunksDB()
-    }
-
-    @Provides
-    @Singleton
-    fun provideGeminiRemoteAPI(): GeminiRemoteAPI {
-        return GeminiRemoteAPI()
-    }
-
-    @Provides
-    @Singleton
-    fun provideSentenceEncoder(context: Application): SentenceEmbeddingProvider {
-        return SentenceEmbeddingProvider(context)
-    }
-}
+@Module @ComponentScan("com.ml.shubham0204.docqa")
+class AppModule

--- a/app/src/main/java/com/ml/shubham0204/docqa/domain/ChunksUseCase.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/domain/ChunksUseCase.kt
@@ -4,13 +4,13 @@ import android.util.Log
 import com.ml.shubham0204.docqa.data.Chunk
 import com.ml.shubham0204.docqa.data.ChunksDB
 import com.ml.shubham0204.docqa.domain.embeddings.SentenceEmbeddingProvider
-import javax.inject.Inject
-import javax.inject.Singleton
+import org.koin.core.annotation.Single
 
-@Singleton
-class ChunksUseCase
-@Inject
-constructor(private val chunksDB: ChunksDB, private val sentenceEncoder: SentenceEmbeddingProvider) {
+@Single
+class ChunksUseCase(
+    private val chunksDB: ChunksDB,
+    private val sentenceEncoder: SentenceEmbeddingProvider
+) {
 
     fun addChunk(docId: Long, docFileName: String, chunkText: String) {
         val embedding = sentenceEncoder.encodeText(chunkText)

--- a/app/src/main/java/com/ml/shubham0204/docqa/domain/DocumentsUseCase.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/domain/DocumentsUseCase.kt
@@ -6,17 +6,17 @@ import com.ml.shubham0204.docqa.data.DocumentsDB
 import com.ml.shubham0204.docqa.domain.readers.Readers
 import com.ml.shubham0204.docqa.domain.splitters.WhiteSpaceSplitter
 import java.io.InputStream
-import javax.inject.Inject
-import javax.inject.Singleton
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.withContext
 import setProgressDialogText
+import org.koin.core.annotation.Single
 
-@Singleton
-class DocumentsUseCase
-@Inject
-constructor(private val chunksUseCase: ChunksUseCase, private val documentsDB: DocumentsDB) {
+@Single
+class DocumentsUseCase(
+    private val chunksUseCase: ChunksUseCase,
+    private val documentsDB: DocumentsDB
+) {
 
     suspend fun addDocument(
         inputStream: InputStream,

--- a/app/src/main/java/com/ml/shubham0204/docqa/domain/QAUseCase.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/domain/QAUseCase.kt
@@ -4,16 +4,13 @@ import android.util.Log
 import com.ml.shubham0204.docqa.data.QueryResult
 import com.ml.shubham0204.docqa.data.RetrievedContext
 import com.ml.shubham0204.docqa.domain.llm.GeminiRemoteAPI
-import javax.inject.Inject
-import javax.inject.Singleton
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
+import org.koin.core.annotation.Single
 
-@Singleton
-class QAUseCase
-@Inject
-constructor(
+@Single
+class QAUseCase(
     private val documentsUseCase: DocumentsUseCase,
     private val chunksUseCase: ChunksUseCase,
     private val geminiRemoteAPI: GeminiRemoteAPI

--- a/app/src/main/java/com/ml/shubham0204/docqa/domain/embeddings/SentenceEmbeddingProvider.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/domain/embeddings/SentenceEmbeddingProvider.kt
@@ -5,7 +5,9 @@ import com.ml.shubham0204.sentence_embeddings.SentenceEmbedding
 import java.io.File
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.runBlocking
+import org.koin.core.annotation.Single
 
+@Single
 class SentenceEmbeddingProvider(private val context: Context) {
 
     private val sentenceEmbedding = SentenceEmbedding()

--- a/app/src/main/java/com/ml/shubham0204/docqa/domain/llm/GeminiRemoteAPI.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/domain/llm/GeminiRemoteAPI.kt
@@ -2,13 +2,14 @@ package com.ml.shubham0204.docqa.domain.llm
 
 import android.util.Log
 import com.google.ai.client.generativeai.GenerativeModel
-import com.google.ai.client.generativeai.type.Content
 import com.google.ai.client.generativeai.type.GenerationConfig
 import com.google.ai.client.generativeai.type.content
 import com.ml.shubham0204.docqa.BuildConfig
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
+import org.koin.core.annotation.Single
 
+@Single
 class GeminiRemoteAPI {
 
     private val apiKey = BuildConfig.geminiKey

--- a/app/src/main/java/com/ml/shubham0204/docqa/ui/screens/ChatScreen.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/ui/screens/ChatScreen.kt
@@ -47,11 +47,11 @@ import androidx.compose.ui.text.TextStyle
 import androidx.compose.ui.text.font.FontStyle
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.sp
-import androidx.hilt.navigation.compose.hiltViewModel
 import com.ml.shubham0204.docqa.R
 import com.ml.shubham0204.docqa.ui.theme.DocQATheme
 import com.ml.shubham0204.docqa.ui.viewmodels.ChatViewModel
 import dev.jeziellago.compose.markdowntext.MarkdownText
+import org.koin.androidx.compose.koinViewModel
 
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
@@ -74,7 +74,7 @@ fun ChatScreen(onOpenDocsClick: (() -> Unit)) {
             }
         ) { innerPadding ->
             Column(modifier = Modifier.padding(innerPadding).padding(16.dp).fillMaxWidth()) {
-                val chatViewModel: ChatViewModel = hiltViewModel()
+                val chatViewModel: ChatViewModel = koinViewModel()
                 Column {
                     QALayout(chatViewModel)
                     Spacer(modifier = Modifier.height(8.dp))

--- a/app/src/main/java/com/ml/shubham0204/docqa/ui/screens/DocsScreen.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/ui/screens/DocsScreen.kt
@@ -50,7 +50,6 @@ import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.text.font.FontWeight
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.window.Dialog
-import androidx.hilt.navigation.compose.hiltViewModel
 import com.ml.shubham0204.docqa.data.Document
 import com.ml.shubham0204.docqa.domain.readers.Readers
 import com.ml.shubham0204.docqa.ui.components.AppAlertDialog
@@ -62,6 +61,7 @@ import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
+import org.koin.androidx.compose.koinViewModel
 import showProgressDialog
 
 private val showDocDetailDialog = mutableStateOf(false)
@@ -92,7 +92,7 @@ fun DocsScreen(onBackClick: (() -> Unit)) {
                 )
             }
         ) { innerPadding ->
-            val docsViewModel: DocsViewModel = hiltViewModel()
+            val docsViewModel: DocsViewModel = koinViewModel()
             Column(modifier = Modifier.padding(innerPadding).padding(16.dp).fillMaxWidth()) {
                 Spacer(modifier = Modifier.height(12.dp))
                 DocsList(docsViewModel)

--- a/app/src/main/java/com/ml/shubham0204/docqa/ui/viewmodels/ChatViewModel.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/ui/viewmodels/ChatViewModel.kt
@@ -4,11 +4,10 @@ import androidx.compose.runtime.mutableStateOf
 import androidx.lifecycle.ViewModel
 import com.ml.shubham0204.docqa.data.RetrievedContext
 import com.ml.shubham0204.docqa.domain.QAUseCase
-import dagger.hilt.android.lifecycle.HiltViewModel
-import javax.inject.Inject
+import org.koin.android.annotation.KoinViewModel
 
-@HiltViewModel
-class ChatViewModel @Inject constructor(val qaUseCase: QAUseCase) : ViewModel() {
+@KoinViewModel
+class ChatViewModel(val qaUseCase: QAUseCase) : ViewModel() {
 
     val questionState = mutableStateOf("")
     val responseState = mutableStateOf("")

--- a/app/src/main/java/com/ml/shubham0204/docqa/ui/viewmodels/DocsViewModel.kt
+++ b/app/src/main/java/com/ml/shubham0204/docqa/ui/viewmodels/DocsViewModel.kt
@@ -2,11 +2,10 @@ package com.ml.shubham0204.docqa.ui.viewmodels
 
 import androidx.lifecycle.ViewModel
 import com.ml.shubham0204.docqa.domain.DocumentsUseCase
-import dagger.hilt.android.lifecycle.HiltViewModel
-import javax.inject.Inject
+import org.koin.android.annotation.KoinViewModel
 
-@HiltViewModel
-class DocsViewModel @Inject constructor(val documentsUseCase: DocumentsUseCase) : ViewModel() {
+@KoinViewModel
+class DocsViewModel(val documentsUseCase: DocumentsUseCase) : ViewModel() {
 
     val documentsFlow = documentsUseCase.getAllDocuments()
 }

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -9,5 +9,5 @@ plugins {
     alias(libs.plugins.android.application) apply false
     alias(libs.plugins.jetbrains.kotlin.android) apply false
     alias(libs.plugins.compose.compiler) apply false
-    id("com.google.dagger.hilt.android") version "2.50" apply false
+    id("com.google.devtools.ksp") version "2.0.0-1.0.24" apply false
 }

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -12,8 +12,8 @@ navigationComposeVersion = "2.7.7"
 objectboxGradlePlugin = "4.0.0"
 pdfboxVersion = "2.0.0"
 apachePoi = "5.2.5"
-hiltAndroid = "2.50"
-hiltNavigationCompose = "1.2.0"
+koin = "3.5.6"
+koinAnnotations = "1.3.1"
 
 [libraries]
 androidx-core-ktx = { group = "androidx.core", name = "core-ktx", version.ref = "coreKtx" }
@@ -31,14 +31,16 @@ androidx-ui-tooling-preview = { group = "androidx.compose.ui", name = "ui-toolin
 androidx-ui-test-manifest = { group = "androidx.compose.ui", name = "ui-test-manifest" }
 androidx-ui-test-junit4 = { group = "androidx.compose.ui", name = "ui-test-junit4" }
 androidx-material3 = { group = "androidx.compose.material3", name = "material3" }
-androidx-hilt-navigation-compose = { module = "androidx.hilt:hilt-navigation-compose", version.ref = "hiltNavigationCompose" }
-apache-pdfbox = { module = "org.apache.pdfbox:pdfbox", version.ref = "pdfboxVersion" }
+
 apache-poi = { module = "org.apache.poi:poi", version.ref = "apachePoi" }
 apache-poi-ooxml = { module = "org.apache.poi:poi-ooxml", version.ref = "apachePoi" }
 navigation-compose = { module = "androidx.navigation:navigation-compose", version.ref = "navigationComposeVersion" }
 objectbox = { module = "io.objectbox:objectbox-gradle-plugin", version.ref = "objectboxGradlePlugin" }
-hilt-android = { module = "com.google.dagger:hilt-android", version.ref = "hiltAndroid" }
-hilt-android-compiler = { module = "com.google.dagger:hilt-android-compiler", version.ref = "hiltAndroid" }
+
+koin-android = { module = "io.insert-koin:koin-android", version.ref = "koin" }
+koin-androidx-compose = { module = "io.insert-koin:koin-androidx-compose", version.ref = "koin" }
+koin-annotations = { module = "io.insert-koin:koin-annotations", version.ref = "koinAnnotations" }
+koin-ksp-compiler = { module = "io.insert-koin:koin-ksp-compiler", version.ref = "koinAnnotations" }
 
 [plugins]
 android-application = { id = "com.android.application", version.ref = "agp" }

